### ISSUE TITLE
Phase 0: Build the verified whole-proof evaluation loop

### ISSUE THREAD

**Issue by murillo128** [murillo128/qwen-lean#3]
# Phase 0: Build the verified whole-proof evaluation loop

## Readiness
**Initial state:** execution-ready

This is the controlling execution issue for Phase 0 of epic #1. It implements the Phase 0 exit gate from `PLAN.md` and must preserve the accepted repository decisions in `docs/DECISIONS.md`.

## Goal and current limitation

Build the smallest trustworthy end-to-end loop that can take a Lean theorem task, produce or accept a candidate whole proof, verify that candidate with Lean, and write compact machine-readable results.

The repository currently has planning and workflow scaffolding but no executable model/evaluator runtime. Until this phase exists, later baseline or fine-tuning results would have no stable task contract or trustworthy verifier path.

Phase 0 is complete when a small fixed fixture set can run repeatedly through the same prompt/reconstruction/verifier contract, known-valid and known-invalid candidates are classified correctly, and `Qwen/Qwen3-8B-Base` can be loaded for a real generation smoke run whose candidate is sent through that exact verifier path.

This phase does **not** try to establish model quality. A Qwen-generated proof may pass or fail; the requirement is that generation and verification are wired together correctly and the result is recorded faithfully.

### Accepted project decisions

- Task: whole-proof generation (`D001`).
- Success: verifier acceptance, not textual resemblance or training loss (`D002`).
- Base model for the first cycle: `Qwen/Qwen3-8B-Base` (`D003`).
- Runtime stack: Python + PyTorch + Transformers; Lean 4/Lake for verification (`D005`).
- The repository is agent-operated; a small command surface is useful, but a polished human CLI/notebook workflow is not a goal (`D006`).
- Large model/cache artifacts stay outside Git; small fixtures and compact evidence may be committed (`D011`).
- Practical comparability is required, not publication-grade reproducibility (`D012`).

### Phase-0 runtime baseline

- Python environment is managed with `uv` and committed lock state.
- Only dependencies required for this phase should be introduced; do not scaffold TRL/PEFT/bitsandbytes training code yet.
- Use direct Transformers generation for the Phase 0 model smoke path. vLLM is explicitly deferred until throughput warrants it.
- **All model inference/generation for this project is executed on project-controlled local GPU compute. For Phase 0, use the available NVIDIA Ada GPU. Do not use Hugging Face Jobs, Hugging Face Inference API/Endpoints, Spaces compute, or another hosted inference/GPU-job service.**
- Hugging Face Hub is allowed as an **artifact source** for downloading/caching `Qwen/Qwen3-8B-Base` and its tokenizer. `from_pretrained(...)` downloading model files is not remote inference; after loading, generation must execute locally on the Ada GPU.
- The implementation must not introduce a hosted-inference client or make successful evaluation depend on remote inference availability or remote compute credits.
- Pin the Phase 0 Lean fixture environment to Lean `v4.32.0` and mathlib `v4.32.0`, with normal Lake dependency locking committed so a fresh executor gets the same verifier environment.
- Standard fixture/unit tests must not require a GPU or downloading the 8B model. The full model smoke is a separate local-GPU validation step.

This local-inference constraint is durable beyond Phase 0. The implementation PR must update `docs/DECISIONS.md` so later phases do not reinterpret provider-neutral artifact/storage choices as permission to run inference on hosted services.

### 1. Task representation

Use a small machine-readable task record with, at minimum:

- `id`: stable task identifier;
- `preamble`: exact Lean source required before the theorem, including imports and any declarations/options/open namespaces needed by the theorem;
- `declaration`: the theorem declaration through its result type, excluding `:= by` and excluding the proof;
- `declaration_name`: the declared theorem name, available to verifier-side checks that need to inspect the resulting declaration.

Do not parse arbitrary mathlib source files in Phase 0. Later dataset/benchmark loaders may produce this task shape from richer sources.

### 2. Prompt/proof serialization (`whole-proof-v1`)

Define one versioned prompt format named `whole-proof-v1`.

For a task, render a Lean source prefix equivalent to:

```lean
<preamble>

/- Complete the proof below.
Return only Lean code continuing after `by`; do not use `sorry` or `admit`. -/
<declaration> := by
  
```

Generation starts after the two-space indentation on the final line.

Requirements:

- The model receives plain code-completion text; do not apply an instruction/chat template to the base checkpoint.
- The decoded generated continuation is the candidate text. Remove tokenizer special tokens, but do **not** strip Markdown fences, prose prefixes, `by`, or other model mistakes to make a candidate pass.
- Normalizing line endings/trailing transport whitespace is allowed; semantic rewriting or proof extraction is not.
- Reconstruct verifier input from the same prompt prefix plus the candidate and a final newline. Do not maintain a second independently formatted theorem template.
- An empty decoded continuation is not a Lean success; record it as an output/generation failure.

This issue resolves the exact prompt/proof-target serialization that `D001` currently leaves `OPEN`. Because later phases depend on this shared contract, the implementation PR must update `D001` to record `whole-proof-v1` as `ACCEPTED` without copying the entire issue into `docs/DECISIONS.md`.

### 3. Candidate sources

The evaluator must support the same downstream verifier path from two candidate sources:

1. **provided fixture candidates** — deterministic text supplied by tests/fixtures;
2. **model candidates** — raw continuations generated locally by `Qwen/Qwen3-8B-Base` on the project Ada GPU.

Do not create separate verifier implementations for these paths.

### 4. Lean verification

For every candidate:

1. render a fresh temporary `.lean` source from the exact task prompt prefix plus candidate;
2. run it in the pinned Lake environment, e.g. through `lake env lean` or an equivalent repository-native invocation;
3. enforce a per-candidate timeout (Phase 0 default: 30 seconds, configurable);
4. capture exit status, elapsed verification time, and useful stdout/stderr diagnostics;
5. return a structured verifier outcome instead of inferring success from process completion alone.

Candidates must be isolated so a malformed proof cannot corrupt evaluation of later tasks.

At minimum use these result categories:

- `verified` — Lean accepted the reconstructed theorem **and** the declaration is free of proof placeholders such as `sorry`/`admit`;
- `lean_rejected` — Lean ran normally and rejected the reconstructed source;
- `empty_candidate` — generation/fixture supplied no proof continuation;
- `verifier_timeout` — the Lean check exceeded its configured limit;
- `verifier_error` — Lean/Lake could not be executed or failed as infrastructure rather than as a normal rejected theorem;
- `generation_error` — model loading/generation failed before a candidate could be verified.

Do not build a brittle parser that tries to classify every Lean diagnostic into syntax vs elaboration vs tactic failure in Phase 0. Preserve the diagnostics so later analysis can add stable subcategories if they become useful.

The `verified` path must include an explicit automated check proving that a candidate containing an unsound placeholder does not receive `verified`. The implementation may choose the cleanest Lean-aware mechanism; plain exit code alone is insufficient.

### 5. Model generation smoke path

Add a direct Transformers path that loads `Qwen/Qwen3-8B-Base` and its matching tokenizer into the local process and generates one candidate per selected task on the project Ada GPU.

Phase 0 smoke defaults:

- greedy generation (`do_sample=false`);
- `max_new_tokens=128`;
- stop normally on the model/tokenizer EOS condition or the token limit;
- use a local CUDA dtype/device configuration that loads the unchanged base checkpoint without applying a LoRA adapter or training quantization path.

`AutoTokenizer.from_pretrained(...)` / `AutoModelForCausalLM.from_pretrained(...)` may fetch artifacts from Hugging Face Hub into the local cache. They must not be replaced by a remote inference API or remote GPU job. The actual forward passes and `generate()` call must run on the local Ada CUDA device.

These generation defaults are only a deterministic Phase 0 plumbing check. Phase 1 will define the sampling settings used for pass@k baseline measurement. Do not present Phase 0 smoke output as a model-quality baseline.

### 6. Machine-readable results

Write compact versioned evaluation artifacts with run-level metadata separated from candidate records so later runs do not repeat large metadata on every line.

A run metadata record/file must capture at least:

- result schema version;
- candidate source (`fixture` or `model`);
- task source/fixture identity;
- prompt format id (`whole-proof-v1`);
- Lean toolchain/mathlib identity;
- for model runs: model id, tokenizer identity, material generation settings, and the local CUDA device identity;
- timeout and other verifier settings that affect interpretation.

Each candidate result must capture at least:

- task id;
- candidate index/id;
- raw candidate text;
- final category;
- Lean exit code when a Lean process ran;
- verifier diagnostics useful for debugging;
- generation latency when applicable;
- verification latency;
- total or otherwise reconstructable end-to-end timing.

JSON + JSONL is the preferred simple shape (`run.json` + `results.jsonl`), but an equally inspectable versioned representation is acceptable if it preserves the same separation and fields.

Do not add hashes, immutable artifact archives, experiment-tracking services, or large logs merely for Phase 0 bookkeeping.

### In scope

- minimal `uv`/Python project runtime needed by the evaluator;
- pinned Lean/Lake/mathlib fixture environment;
- task record and exact `whole-proof-v1` renderer;
- shared candidate verification path;
- deterministic fixture-candidate evaluation;
- direct local Transformers Qwen3-8B-Base generation on the project Ada GPU;
- compact structured result artifacts;
- unit/integration tests and a small fixed Lean fixture set;
- small agent-friendly commands/module entry points to run deterministic evaluation and the local-GPU smoke;
- the minimal durable decision-log updates described above.

### Out of scope

- miniF2F or another external benchmark;
- pass@k baseline measurement or claims about Qwen quality;
- training datasets or proof extraction from mathlib;
- SFT, QLoRA, PEFT, bitsandbytes training, TRL trainers, or checkpoints;
- vLLM integration;
- Hugging Face Hub artifact publication;
- **hosted inference or hosted GPU execution, including Hugging Face Jobs, Hugging Face Inference API/Endpoints, Spaces compute, or equivalent services**;
- synthetic data, self-training, GRPO/RL, tactic-level proving, retrieval, or search;
- elaborate experiment tracking or reproducibility infrastructure;
- exhaustive Lean diagnostic taxonomy.

### Invariants

- `Qwen/Qwen3-8B-Base` remains the unchanged base model for this phase.
- All model generation/inference executes locally on the project Ada GPU; remote services may distribute model artifacts but may not execute the model.
- Fixture candidates and model candidates use the same prompt reconstruction and Lean verifier implementation.
- `verified` always means the declared Lean environment accepted the reconstructed theorem without proof placeholders.
- No post-processing may silently repair a candidate before verification.
- Model/tokenizer identity, prompt format, Lean environment, or inference execution mode must not change silently within a run.
- A model-generation failure, Lean rejection, timeout, and verifier infrastructure error remain distinguishable.
- Standard test execution remains possible without GPU/model download; the local-GPU smoke is explicit.

### A. Repository tests — required

Add automated tests that prove at least:

1. exact `whole-proof-v1` rendering for a known task;
2. exact source reconstruction uses that same prefix plus raw candidate text;
3. structured result serialization/deserialization or schema validation;
4. known-valid Lean candidates return `verified`;
5. a syntax/elaboration/proof failure returns `lean_rejected` with diagnostics;
6. an empty candidate returns `empty_candidate` without being sent through a misleading success path;
7. a `sorry`/`admit` placeholder candidate does **not** return `verified`;
8. one failing fixture does not prevent later fixtures from being evaluated.

Use a small fixed fixture set with at least:

- one Core-Lean-style theorem such as hypothesis/identity reuse;
- one mathlib-backed theorem/tactic so the pinned dependency path is exercised;
- negative candidates covering rejection and placeholder handling.

The normal test command must run through `uv` and must not require a GPU or the Qwen checkpoint.

### B. Deterministic end-to-end fixture run — required

Provide one repository command/module invocation that evaluates the fixed provided candidates through the production evaluator and writes the structured artifacts.

Pass gate:

- every fixture produces one expected result record;
- expected valid candidates are `verified`;
- expected invalid and placeholder candidates are not `verified` and have the correct top-level category;
- artifacts contain the required run and candidate metadata;
- rerunning the same fixture workload does not change prompt format or expected classifications.

The exact wall-clock timings may differ between runs and are not part of the equality gate.

### C. Real local-Ada Qwen generation smoke — required

Run at least one fixed Phase 0 task through the actual `Qwen/Qwen3-8B-Base` local loading/generation path on the project Ada GPU and then through the same Lean verifier.

Pass gate:

- model and tokenizer artifacts load successfully into the local runtime;
- metadata identifies the local CUDA/Ada device used for inference;
- a raw candidate is generated or a generation failure is represented explicitly;
- any generated candidate is verified through the production Lean path without repair/extraction;
- structured output records model/tokenizer identity, generation settings, candidate text, verifier outcome, and timings;
- no Hugging Face Job, hosted endpoint, hosted inference call, or remote GPU execution is involved;
- the run makes no claim that a proof success/failure is a baseline quality result.

Record only compact evidence needed to inspect that the path worked; do not commit model caches or bulky logs.

### Final-capable evaluator-contract review

After implementation and all validation above, run one independent review over the complete Phase 0 diff and compact evidence.

The reviewer must specifically inspect:

- whether `whole-proof-v1` is rendered identically for generation and verification;
- whether any post-processing can accidentally turn invalid model text into a passing proof;
- whether `sorry`/`admit` can be miscounted as `verified`;
- whether normal Lean rejection is separated from infrastructure failure/timeout;
- whether the GPU smoke actually uses `Qwen/Qwen3-8B-Base` on the project Ada GPU and the same evaluator;
- whether the implementation contains no hosted-inference/hosted-GPU execution dependency;
- whether the durable decision-log update matches the local-inference contract;
- whether the `D001` decision-log update matches the implemented prompt contract;
- whether the Phase 0 exit gate from `PLAN.md` is objectively satisfied.

Progression is unsafe if the evaluator can produce a false-positive `verified`, if fixture and model paths use different semantics, if inference runs on a hosted service, or if the prompt contract is not explicit and stable.

This checkpoint is **final-capable** if it covers the complete final diff and all remaining acceptance criteria.

## Delivery

Deliver Phase 0 as one focused PR linked to this issue unless implementation reveals a genuinely independent blocker that warrants a separate controlling issue.

The PR should contain the runtime/evaluator code, pinned small environment/configuration files, fixed fixtures/tests, the durable decision updates, and only compact evidence from the deterministic and local-Ada smoke runs.

Do not commit model weights, caches, large datasets, or bulky runtime logs.

On successful final review and merge, this issue can move to `completed`; epic #1 can then mark Phase 0 complete and make Phase 1 the current phase.

**Comment by murillo128** [comment]
## Blocked

**Delivered or confirmed:**
- Phase 0 runtime, pinned Lean/mathlib and `uv` environments, fixtures, tests, and deterministic evidence are published in draft PR #4.
- `uv run pytest` passes all 11 tests; two production fixture runs preserve `whole-proof-v1` and the expected classifications.

**Validation:** The required Hugging Face `a10g-large` GPU job was submitted against the published implementation, but the Jobs API rejected it before scheduling with `402 Payment Required`. This runner has no CUDA device.

**Material issue:** The required real `Qwen/Qwen3-8B-Base` GPU smoke and therefore the final-capable independent review cannot be completed without GPU compute.

**Next:** Add a positive Hugging Face Jobs credit balance for the connected `murillo2000` account, or provide another CUDA runner with at least 24 GB VRAM, then rerun the documented `model-smoke` command and retain its compact JSON evidence.

**Comment by murillo128** [comment]
## Contract amendment — local inference

The prior blocker based on Hugging Face Jobs credit is superseded. Hosted GPU execution is not part of the project contract.

- All model generation/inference must run on project-controlled local GPU compute; for Phase 0, use the available NVIDIA Ada GPU.
- Hugging Face Hub may be used only to download/cache model and tokenizer artifacts.
- Do not use Hugging Face Jobs, Inference API/Endpoints, Spaces compute, or equivalent hosted inference/GPU-job services.
- The Phase 0 GPU smoke must be rerun on the local Ada and record the CUDA device identity in compact evidence.
- Update `docs/DECISIONS.md` in PR #4 so this remains a durable project-wide constraint.

Issue state has been changed from `blocked` to `in-progress`; inability to purchase hosted compute is not a blocker.

**Comment by murillo128** [comment]
## Completed

- PR #4 merged to `main` as `87cd450af873083c0589fca57c1a6c838fdb770c`.
- Repository tests pass 11/11 and repeat fixture runs preserve the expected classifications.
- The real `Qwen/Qwen3-8B-Base` smoke ran offline from the local artifact cache on `NVIDIA RTX 4000 Ada Generation`; its raw candidate passed through the shared Lean verifier and compact evidence is committed.
- Final-capable independent review returned `PASS_WITH_NOTES` on exact target `4dc28bc0d8e6d57ca097d037b996a2030034a19e`, with no material finding.
- No hosted inference or hosted GPU execution was used or introduced.

The Phase 0 exit gate is satisfied. The reviewer note about fixture-source metadata for a future non-default `--fixtures` invocation is non-blocking and does not affect this fixed Phase 0 smoke.
